tests/serial/test_serialize.py

```diff
@@ -136,6 +139,127 @@ def test_raises_not_implemented_error_for_invalid_combination(
+        @dataclass
+        class E:
+            client_id: int
+
+        [field] = fields(E)
+        writer = get_field_writer(
+            field=field, flexible=True, is_request_header=True, is_tag=False
+        )
+        assert writer == writers.write_nullable_legacy_string
+
+    def test_returns_plain_writer_for_primitive_field(self) -> None:
+        @dataclass
+        class E:
+            a: i8 = field(metadata={"kafka_type": "int8"})
+
+        [extracted_field] = fields(E)
+        writer = get_field_writer(
+            field=extracted_field, flexible=True, is_request_header=True, is_tag=False
+        )
+        assert writer == writers.write_int8
+
+    def test_returns_array_writer_for_primitive_tuple_field(
+        self,
+        buffer: io.BytesIO,
+    ) -> None:
+        @dataclass
+        class E:
+            a: tuple[i8, ...] = field(metadata={"kafka_type": "int8"})
+
+        [extracted_field] = fields(E)
+        writer = get_field_writer(
+            field=extracted_field, flexible=True, is_request_header=True, is_tag=False
+        )
+
+        # We test that the returned object is an int8 array writer, by making
+        # sure it behaves like one.
+        writer(buffer, [1, 2, 3])
+        buffer.seek(0)
+        assert read_compact_array_length(buffer) == 3
+        assert read_int8(buffer) == 1
+        assert read_int8(buffer) == 2
+        assert read_int8(buffer) == 3
+
+    def test_returns_entity_writer_for_entity_field(
+        self,
+        buffer: io.BytesIO,
+    ) -> None:
+        @dataclass
+        class A:
+            __flexible__: ClassVar = True
+            f: i8 = field(metadata={"kafka_type": "int8"})
+
+        @dataclass
+        class B:
+            a: A
+
+        [extracted_field] = fields(B)
+        writer = get_field_writer(
+            field=extracted_field, flexible=True, is_request_header=True, is_tag=False
+        )
+
+        writer(buffer, A(f=i8(23)))
+        buffer.seek(0)
+        assert read_int8(buffer) == 23
+        assert read_unsigned_varint(buffer) == 0  # tags
+
+    def test_returns_entity_writer_for_nullable_entity_field(
+        self,
+        buffer: io.BytesIO,
+    ) -> None:
+        @dataclass
+        class A:
+            __flexible__: ClassVar = True
+            f: i8 = field(metadata={"kafka_type": "int8"})
+
+        @dataclass
+        class B:
+            a: A | None
+
+        [extracted_field] = fields(B)
+        writer = get_field_writer(
+            field=extracted_field, flexible=True, is_request_header=True, is_tag=False
+        )
+
+        writer(buffer, A(f=i8(23)))
+        writer(buffer, None)
+        buffer.seek(0)
+        assert NullableEntityMarker(read_int8(buffer)) is NullableEntityMarker.not_null
+        assert read_int8(buffer) == 23
+        assert read_unsigned_varint(buffer) == 0  # tags
+        assert NullableEntityMarker(read_int8(buffer)) is NullableEntityMarker.null
```

Comment:
Just want to make sure i understand it: Why is there a Nullable marker at the end? In the code it looks like it is only written at the beginning of the entity?

Reply:
ah forget it, you write None before. Didn't see that